File: JAVA/Pertemuan 01/src/id/ac/uph/pbo/pertemuan01/OOPSolusi.java
package id.ac.uph.pbo.pertemuan01;

public class OOPSolusi {

    public static void main(String[] args) {

        System.out.println("╔══════════════════════════════════════════════╗");
        System.out.println("║   PENDEKATAN OOP – Solusi dengan Enkapsulasi ║");
        System.out.println("╚══════════════════════════════════════════════╝\n");

        // ── Membuat objek dari class Mahasiswa ───────────────────────────────
        Mahasiswa mhs1 = new Mahasiswa("Budi Santoso",  "220301001", 3.50);
        Mahasiswa mhs2 = new Mahasiswa("Sari Wulandari","220301002", 3.75);
        Mahasiswa mhs3 = new Mahasiswa("Andi Pratama",  "220301003", 2.90);

        System.out.println("Data Mahasiswa yang baru dibuat:");
        mhs1.tampilkanInfo();
        mhs2.tampilkanInfo();
        mhs3.tampilkanInfo();

        // ── Demonstrasi method perilaku ──────────────────────────────────────
        System.out.println();
        mhs1.belajar();
        mhs2.hadirKuliah("Pemrograman Berorientasi Objek");

        // ── Mencoba mengubah IPK dengan nilai TIDAK VALID ────────────────────
        System.out.println("\n── Mencoba mengubah IPK Andi ke -99.0 (tidak valid): ──");
        mhs3.setIpk(-99.0);           // ditolak oleh validasi setter
        System.out.print("  IPK Andi setelah percobaan: ");
        System.out.printf("%.2f (tidak berubah)%n", mhs3.getIpk());

        // ── Mengubah IPK dengan nilai VALID ─────────────────────────────────
        System.out.println("\n── Mengubah IPK Andi ke 3.10 (valid): ──");
        mhs3.setIpk(3.10);
        System.out.print("  IPK Andi sekarang: ");
        System.out.printf("%.2f%n", mhs3.getIpk());

        System.out.println("\n── Data akhir semua mahasiswa: ──");
        mhs1.tampilkanInfo();
        mhs2.tampilkanInfo();
        mhs3.tampilkanInfo();

        System.out.println("\n──────────────────────────────────────────────────");
        System.out.println("KESIMPULAN: Enkapsulasi melindungi data IPK lewat");
        System.out.println("           setter yang tervalidasi – data aman!");
        System.out.println("──────────────────────────────────────────────────");
    }
}

File: PYTHON/Pertemuan 01/OOPSolusi.py
import sys
sys.stdout.reconfigure(encoding='utf-8')
from Mahasiswa import Mahasiswa

def main():
    print("╔══════════════════════════════════════════════╗")
    print("║   PENDEKATAN OOP – Solusi dengan Enkapsulasi ║")
    print("╚══════════════════════════════════════════════╝\n")

    # ── Membuat objek dari class Mahasiswa ───────────────────────────────
    mhs1 = Mahasiswa("Budi Santoso",  "220301001", 3.50)
    mhs2 = Mahasiswa("Sari Wulandari","220301002", 3.75)
    mhs3 = Mahasiswa("Andi Pratama",  "220301003", 2.90)

    print("Data Mahasiswa yang baru dibuat:")
    mhs1.tampilkan_info()
    mhs2.tampilkan_info()
    mhs3.tampilkan_info()

    # ── Demonstrasi method perilaku ──────────────────────────────────────
    print()
    mhs1.belajar()
    mhs2.hadir_kuliah("Pemrograman Berorientasi Objek")

    # ── Mencoba mengubah IPK dengan nilai TIDAK VALID ────────────────────
    print("\n── Mencoba mengubah IPK Andi ke -99.0 (tidak valid): ──")
    mhs3.ipk = -99.0           # ditolak oleh validasi setter
    print("  IPK Andi setelah percobaan: ", end="")
    print(f"{mhs3.ipk:.2f} (tidak berubah)")

    # ── Mengubah IPK dengan nilai VALID ─────────────────────────────────
    print("\n── Mengubah IPK Andi ke 3.10 (valid): ──")
    mhs3.ipk = 3.10
    print("  IPK Andi sekarang: ", end="")
    print(f"{mhs3.ipk:.2f}")

    print("\n── Data akhir semua mahasiswa: ──")
    mhs1.tampilkan_info()
    mhs2.tampilkan_info()
    mhs3.tampilkan_info()

    print("\n──────────────────────────────────────────────────")
    print("KESIMPULAN: Enkapsulasi melindungi data IPK lewat")
    print("           setter yang tervalidasi – data aman!")
    print("──────────────────────────────────────────────────")

if __name__ == "__main__":
    main()

